pkg/applicationserver/applicationserver.go

```diff
@@ -1055,3 +1059,33 @@ func (as *ApplicationServer) GetMQTTConfig(ctx context.Context) (*config.MQTT, e
 	}
 	return &config.MQTT, nil
 }
+
+func (as *ApplicationServer) getCachedEndDevice(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (*ttnpb.EndDevice, error) {
```

Comment:
As discussed in the original issue - this is not ok as I cannot actually specify the field mask, and the function signature doesn't mention which fields I get.

Consider looking at the single flight and cached fetcher in https://github.com/TheThingsIndustries/lorawan-stack/blob/v3.10/pkg/tenant/fetcher.tti.go

Reply:
Changed the implementation to an `EndDeviceFetcher`, similar to `TenantFetcher`.